### ISSUE TITLE
Exporting LLama2-7B isn't  working. 

### ISSUE THREAD

**Issue by Anindyadeep** [huggingface/optimum#1606]
### System Info

```shell
Here is the info on GPU drivers and cuda:



+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 545.23.08              Driver Version: 545.23.08    CUDA Version: 12.3     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA A100-SXM4-80GB          Off | 00000000:00:05.0 Off |                    0 |
| N/A   27C    P0              56W / 400W |    139MiB / 81920MiB |      0%      Default |
|                                         |                      |             Disabled |
+-----------------------------------------+----------------------+----------------------+
                                                                                         
+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|    0   N/A  N/A      1156      G   /usr/lib/xorg/Xorg                           73MiB |
|    0   N/A  N/A      1950      G   /usr/bin/gnome-shell                         36MiB |
|    0   N/A  N/A      2529      G   /usr/libexec/gnome-initial-setup              6MiB |
+---------------------------------------------------------------------------------------+


nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2023 NVIDIA Corporation
Built on Fri_Nov__3_17:16:49_PDT_2023
Cuda compilation tools, release 12.3, V12.3.103
Build cuda_12.3.r12.3/compiler.33492891_0
```

### Here are the library versions:

```
torch: '2.1.2+cu121'
transformers: '4.36.1'
onnx: 1.15.0
```

### What all I install?

```
torch
optimum[exporters]
optimum[onnxruntime]
optimum
```
```


### Who can help?

@JingyaHuang @echarlaix 

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction (minimal, reproducible, runnable)

I ran this script:

```bash
optimum-cli export onnx \
--model "$LLAMA_HF_MODEL_DIR" --task text-generation-with-past --framework pt \
--opset 17 --sequence_length 1024 --batch_size 1 --device cuda --fp16 \
"$LLAMA_ONNX_MODEL_DIR" > /dev/null

```

Where inside `LLAMA_HF_MODEL_DIR` includes the following file:

```
models/llama-2-7b-hf
├── config.json
├── generation_config.json
├── model-00001-of-00002.safetensors
├── model-00002-of-00002.safetensors
├── model.safetensors.index.json
├── pytorch_model-00001-of-00002.bin
├── pytorch_model-00002-of-00002.bin
├── pytorch_model.bin.index.json
├── special_tokens_map.json
├── tokenizer_config.json
├── tokenizer.json
└── tokenizer.model
```

Running the above produced the following output:

```
Virtual environment '/home/paperspace/workspace/benchmarks/bench_onnxruntime/venv' created.
Loading checkpoint shards: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:02<00:00,  1.17s/it]
Using the export variant default. Available variants are:
    - default: The default ONNX variant.
Using framework PyTorch: 2.1.2+cu121
Overriding 1 configuration item(s)
        - use_cache -> True
/home/paperspace/workspace/benchmarks/bench_onnxruntime/venv/lib/python3.9/site-packages/transformers/modeling_attn_mask_utils.py:114: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  if (input_shape[-1] > 1 or self.sliding_window is not None) and self.is_causal:
/home/paperspace/workspace/benchmarks/bench_onnxruntime/venv/lib/python3.9/site-packages/optimum/exporters/onnx/model_patcher.py:299: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  if past_key_values_length > 0:
/home/paperspace/workspace/benchmarks/bench_onnxruntime/venv/lib/python3.9/site-packages/transformers/models/llama/modeling_llama.py:150: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  if seq_len > self.max_seq_len_cached:
/home/paperspace/workspace/benchmarks/bench_onnxruntime/venv/lib/python3.9/site-packages/transformers/models/llama/modeling_llama.py:714: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  if attention_mask.size() != (bsz, 1, q_len, kv_seq_len):
In-place op on output of tensor.shape. See https://pytorch.org/docs/master/onnx.html#avoid-inplace-operations-when-using-tensor-shape-in-tracing-mode
In-place op on output of tensor.shape. See https://pytorch.org/docs/master/onnx.html#avoid-inplace-operations-when-using-tensor-shape-in-tracing-mode
In-place op on output of tensor.shape. See https://pytorch.org/docs/master/onnx.html#avoid-inplace-operations-when-using-tensor-shape-in-tracing-mode
In-place op on output of tensor.shape. See https://pytorch.org/docs/master/onnx.html#avoid-inplace-operations-when-using-tensor-shape-in-tracing-mode
In-place op on output of tensor.shape. See https://pytorch.org/docs/master/onnx.html#avoid-inplace-operations-when-using-tensor-shape-in-tracing-mode
In-place op on output of tensor.shape. See https://pytorch.org/docs/master/onnx.html#avoid-inplace-operations-when-using-tensor-shape-in-tracing-mode
In-place op on output of tensor.shape. See https://pytorch.org/docs/master/onnx.html#avoid-inplace-operations-when-using-tensor-shape-in-tracing-mode
In-place op on output of tensor.shape. See https://pytorch.org/docs/master/onnx.html#avoid-inplace-operations-when-using-tensor-shape-in-tracing-mode
In-place op on output of tensor.shape. See https://pytorch.org/docs/master/onnx.html#avoid-inplace-operations-when-using-tensor-shape-in-tracing-mode
In-place op on output of tensor.shape. See https://pytorch.org/docs/master/onnx.html#avoid-inplace-operations-when-using-tensor-shape-in-tracing-mode
In-place op on output of tensor.shape. See https://pytorch.org/docs/master/onnx.html#avoid-inplace-operations-when-using-tensor-shape-in-tracing-mode
In-place op on output of tensor.shape. See https://pytorch.org/docs/master/onnx.html#avoid-inplace-operations-when-using-tensor-shape-in-tracing-mode
In-place op on output of tensor.shape. See https://pytorch.org/docs/master/onnx.html#avoid-inplace-operations-when-using-tensor-shape-in-tracing-mode
In-place op on output of tensor.shape. See https://pytorch.org/docs/master/onnx.html#avoid-inplace-operations-when-using-tensor-shape-in-tracing-mode
In-place op on output of tensor.shape. See https://pytorch.org/docs/master/onnx.html#avoid-inplace-operations-when-using-tensor-shape-in-tracing-mode
In-place op on output of tensor.shape. See https://pytorch.org/docs/master/onnx.html#avoid-inplace-operations-when-using-tensor-shape-in-tracing-mode
In-place op on output of tensor.shape. See https://pytorch.org/docs/master/onnx.html#avoid-inplace-operations-when-using-tensor-shape-in-tracing-mode
In-place op on output of tensor.shape. See https://pytorch.org/docs/master/onnx.html#avoid-inplace-operations-when-using-tensor-shape-in-tracing-mode
In-place op on output of tensor.shape. See https://pytorch.org/docs/master/onnx.html#avoid-inplace-operations-when-using-tensor-shape-in-tracing-mode
In-place op on output of tensor.shape. See https://pytorch.org/docs/master/onnx.html#avoid-inplace-operations-when-using-tensor-shape-in-tracing-mode
In-place op on output of tensor.shape. See https://pytorch.org/docs/master/onnx.html#avoid-inplace-operations-when-using-tensor-shape-in-tracing-mode
In-place op on output of tensor.shape. See https://pytorch.org/docs/master/onnx.html#avoid-inplace-operations-when-using-tensor-shape-in-tracing-mode
In-place op on output of tensor.shape. See https://pytorch.org/docs/master/onnx.html#avoid-inplace-operations-when-using-tensor-shape-in-tracing-mode
In-place op on output of tensor.shape. See https://pytorch.org/docs/master/onnx.html#avoid-inplace-operations-when-using-tensor-shape-in-tracing-mode
In-place op on output of tensor.shape. See https://pytorch.org/docs/master/onnx.html#avoid-inplace-operations-when-using-tensor-shape-in-tracing-mode
In-place op on output of tensor.shape. See https://pytorch.org/docs/master/onnx.html#avoid-inplace-operations-when-using-tensor-shape-in-tracing-mode
In-place op on output of tensor.shape. See https://pytorch.org/docs/master/onnx.html#avoid-inplace-operations-when-using-tensor-shape-in-tracing-mode
In-place op on output of tensor.shape. See https://pytorch.org/docs/master/onnx.html#avoid-inplace-operations-when-using-tensor-shape-in-tracing-mode
In-place op on output of tensor.shape. See https://pytorch.org/docs/master/onnx.html#avoid-inplace-operations-when-using-tensor-shape-in-tracing-mode
In-place op on output of tensor.shape. See https://pytorch.org/docs/master/onnx.html#avoid-inplace-operations-when-using-tensor-shape-in-tracing-mode
In-place op on output of tensor.shape. See https://pytorch.org/docs/master/onnx.html#avoid-inplace-operations-when-using-tensor-shape-in-tracing-mode
In-place op on output of tensor.shape. See https://pytorch.org/docs/master/onnx.html#avoid-inplace-operations-when-using-tensor-shape-in-tracing-mode
Traceback (most recent call last):
  File "/home/paperspace/workspace/benchmarks/bench_onnxruntime/venv/bin/optimum-cli", line 8, in <module>
    sys.exit(main())
  File "/home/paperspace/workspace/benchmarks/bench_onnxruntime/venv/lib/python3.9/site-packages/optimum/commands/optimum_cli.py", line 163, in main
    service.run()
  File "/home/paperspace/workspace/benchmarks/bench_onnxruntime/venv/lib/python3.9/site-packages/optimum/commands/export/onnx.py", line 246, in run
    main_export(
  File "/home/paperspace/workspace/benchmarks/bench_onnxruntime/venv/lib/python3.9/site-packages/optimum/exporters/onnx/__main__.py", line 551, in main_export
    _, onnx_outputs = export_models(
  File "/home/paperspace/workspace/benchmarks/bench_onnxruntime/venv/lib/python3.9/site-packages/optimum/exporters/onnx/convert.py", line 753, in export_models
    export(
  File "/home/paperspace/workspace/benchmarks/bench_onnxruntime/venv/lib/python3.9/site-packages/optimum/exporters/onnx/convert.py", line 856, in export
    export_output = export_pytorch(
  File "/home/paperspace/workspace/benchmarks/bench_onnxruntime/venv/lib/python3.9/site-packages/optimum/exporters/onnx/convert.py", line 573, in export_pytorch
    onnx_export(
  File "/home/paperspace/workspace/benchmarks/bench_onnxruntime/venv/lib/python3.9/site-packages/torch/onnx/utils.py", line 516, in export
    _export(
  File "/home/paperspace/workspace/benchmarks/bench_onnxruntime/venv/lib/python3.9/site-packages/torch/onnx/utils.py", line 1596, in _export
    graph, params_dict, torch_out = _model_to_graph(
  File "/home/paperspace/workspace/benchmarks/bench_onnxruntime/venv/lib/python3.9/site-packages/torch/onnx/utils.py", line 1139, in _model_to_graph
    graph = _optimize_graph(
  File "/home/paperspace/workspace/benchmarks/bench_onnxruntime/venv/lib/python3.9/site-packages/torch/onnx/utils.py", line 677, in _optimize_graph
    graph = _C._jit_pass_onnx(graph, operator_export_type)
  File "/home/paperspace/workspace/benchmarks/bench_onnxruntime/venv/lib/python3.9/site-packages/torch/onnx/utils.py", line 1940, in _run_symbolic_function
    return symbolic_fn(graph_context, *inputs, **attrs)
  File "/home/paperspace/workspace/benchmarks/bench_onnxruntime/venv/lib/python3.9/site-packages/torch/onnx/symbolic_helper.py", line 306, in wrapper
    return fn(g, *args, **kwargs)
  File "/home/paperspace/workspace/benchmarks/bench_onnxruntime/venv/lib/python3.9/site-packages/torch/onnx/symbolic_opset14.py", line 197, in scaled_dot_product_attention
    raise ValueError(
ValueError: Unsupported type for attn_mask: 5
```

### Expected behavior

Although it did make a folder and inside that the following files were there:

```
models/llama-2-7b-onnx/
├── config.json
├── generation_config.json
├── special_tokens_map.json
├── tokenizer_config.json
├── tokenizer.json
└── tokenizer.model
```

Expected behaviour was along with those files, there should be an onnx file present.

**Comment by Anindyadeep** [comment]
Even I ran this command:

```
optimum-cli export onnx --model daryl149/llama-2-7b-chat-hf --device cuda --fp16 --no-post-process llama2_onnx
```

As mentioned in issue #1248 

However, I got the same error, as mentioned above.

**Comment by claeyzre** [comment]
I've got the same problem with some models (TinyLlama-1.1B) and the recent updates, but works with previous version of ORT and transformers and PT

**Comment by Anindyadeep** [comment]
> I've got the same problem with some models (TinyLlama-1.1B) and the recent updates, but works with previous version of ORT and transformers and PT

Hey can you please mention the specifics for the versions you were using in which it works.

**Comment by fxmarty** [comment]
Thank you @Anindyadeep @claeyzre this is a missing feature in PyTorch, which as of 2.1 specifically can not export to ONNX the scaled_dot_product_attention operator to ONNX on CUDA device using fp16, see: https://github.com/pytorch/pytorch/issues/109336

transformers>=4.36 now by default uses SDPA, hence the issue.

I checked, this is indeed fixed using pytorch nightly (hopefully included in pytorch 2.2).

**Comment by Anindyadeep** [comment]
I see, thank you @fxmarty for the update. Will try with 2.2.
